ext/node/ops/sqlite/database.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2018-2025 the Deno authors. MIT license.
+
+use std::cell::Cell;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use deno_core::anyhow;
+use deno_core::anyhow::anyhow;
+use deno_core::op2;
+use deno_core::GarbageCollected;
+use serde::Deserialize;
+
+use super::StatementSync;
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DatabaseSyncOptions {
+  #[serde(default = "true_fn")]
+  open: bool,
+  #[serde(default = "true_fn")]
+  enable_foreign_key_constraints: bool,
+}
+
+fn true_fn() -> bool {
+  true
+}
+
+impl Default for DatabaseSyncOptions {
+  fn default() -> Self {
+    DatabaseSyncOptions {
+      open: true,
+      enable_foreign_key_constraints: true,
+    }
+  }
+}
+
+pub struct DatabaseSync {
+  conn: Rc<RefCell<Option<rusqlite::Connection>>>,
+  options: DatabaseSyncOptions,
+  location: String,
+}
+
+impl GarbageCollected for DatabaseSync {}
+
+// Represents a single connection to a SQLite database.
+#[op2]
+impl DatabaseSync {
+  // Constructs a new `DatabaseSync` instance.
+  //
+  // A SQLite database can be stored in a file or in memory. To
+  // use a file-backed database, the `location` should be a path.
+  // To use an in-memory database, the `location` should be special
+  // name ":memory:".
+  #[constructor]
+  #[cppgc]
+  fn new(
+    #[string] location: String,
+    #[serde] options: Option<DatabaseSyncOptions>,
+  ) -> Result<DatabaseSync, anyhow::Error> {
+    let options = options.unwrap_or_default();
+
+    let db = if options.open {
+      let db = rusqlite::Connection::open(&location)?;
+      if options.enable_foreign_key_constraints {
+        db.execute("PRAGMA foreign_keys = ON", [])?;
+      }
+      Some(db)
+    } else {
+      None
+    };
+
+    Ok(DatabaseSync {
+      conn: Rc::new(RefCell::new(db)),
+      location,
+      options,
+    })
+  }
+
+  // Opens the database specified by `location` of this instance.
+  //
+  // This method should only be used when the database is not opened
+  // via the constructor. An exception is thrown if the database is
+  // already opened.
+  #[fast]
+  fn open(&self) -> Result<(), anyhow::Error> {
+    if self.conn.borrow().is_some() {
+      return Err(anyhow!("Database is already open"));
+    }
+
+    let db = rusqlite::Connection::open(&self.location)?;
```

Comment:
looks like this version doesn't check if deno process has permission to open file at the given location. Can we add the perm check?